packages/flutter/test/material/carousel_test.dart

```diff
@@ -1977,6 +1977,244 @@ void main() {
     });
   });
 
+  group('CarouselView onIndexChanged callback', () {
+    testWidgets('CarouselView shows correct item after animation', (WidgetTester tester) async {
+      final controller = CarouselController();
+      addTearDown(controller.dispose);
+      var leadingIndex = 0;
+
+      await tester.pumpWidget(
+        MaterialApp(
+          home: Scaffold(
+            body: CarouselView(
+              itemExtent: 300,
+              controller: controller,
+              onIndexChanged: (int index) {
+                leadingIndex = index;
+              },
+              children: List<Widget>.generate(6, (int i) => Text('Item $i')),
+            ),
+          ),
+        ),
+      );
+      await tester.pumpAndSettle();
+
+      controller.animateToItem(
+        3,
+        duration: const Duration(milliseconds: 200),
+        curve: Curves.linear,
+      );
+      await tester.pumpAndSettle();
+
+      expect(controller.leadingItem, equals(3));
+      expect(leadingIndex, equals(3));
+
+      controller.animateToItem(
+        1,
+        duration: const Duration(milliseconds: 200),
+        curve: Curves.linear,
+      );
+
+      await tester.pumpAndSettle();
+      expect(controller.leadingItem, equals(1));
+      expect(leadingIndex, equals(1));
+    });
+
+    testWidgets(
+      'CarouselView.weighted shows correct item after animation with symmetric flexWeights',
+      (WidgetTester tester) async {
+        final controller = CarouselController();
+        addTearDown(controller.dispose);
+        var leadingIndex = 0;
+
+        await tester.pumpWidget(
+          MaterialApp(
+            home: Scaffold(
+              body: CarouselView.weighted(
+                flexWeights: const [2, 5, 2],
+                controller: controller,
+                onIndexChanged: (int index) => leadingIndex = index,
+                children: List.generate(6, (i) => Text('Item $i')),
+              ),
+            ),
+          ),
+        );
+        await tester.pumpAndSettle();
+
+        controller.animateToItem(
+          4,
+          duration: const Duration(milliseconds: 200),
+          curve: Curves.linear,
+        );
+        await tester.pumpAndSettle();
+
+        expect(controller.leadingItem, equals(4));
+        expect(leadingIndex, equals(4));
+        expect(find.text('Item 4'), findsOneWidget);
+
+        final Rect viewport = viewportRect(tester);
+        final Rect rect4 = tester.getRect(getItem(4));
+
+        expect(rect4.top, greaterThanOrEqualTo(viewport.top));
+        expect(rect4.bottom, lessThanOrEqualTo(viewport.bottom));
```

Comment:
We either remove this part, or we give precise number for rect4 in the expectation instead of using `greaterThanOrEqualTo`. If [2, 5, 2] causes difficulty to calculate (because the viewport width is 800), we can change it to [2, 4, 2], so that the width of rect 4 should be 200 (if no padding surrounding).